### PR TITLE
Introduce property-based semtype testing

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
This PR does not implement failing test reduction.

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/tests/random.bal [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import ballerina/time;
+import ballerina/test;
+
+// Mask operand of multiplication to 31 bits to avoid overflow.
+isolated class Random {
```
Comment (jclark) [inline compiler/tests/random.bal]: I suggest you put this in the comm.lib module.

File: compiler/modules/front/resolveTypes.bal [hunk 2]
```diff
@@ -39,6 +39,13 @@ function resolveTypes(ModuleSymbols mod) returns ResolveTypeError? {
     }
 }
 
+function resolveType(ModuleSymbols mod, string name) returns ResolveTypeError? {
+    s:ModuleLevelDefn? defn = mod.defns[name];
```
Comment (jclark) [inline compiler/modules/front/resolveTypes.bal]: This isn't doing what resolveTypes is doing...

I suggest you have a `resolveDefn(ModuleLevelDefn function)` which `resolveTypes` calls.

File: compiler/tests/prepositionGenerator.bal [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+
+type PropositionListCtor function () returns Proposition[];
+
+type OnFail function(Context cx, Proposition failedProp);
+
+class Context {
+    final t:Context typeContext;
+    final SubtypeProposition[][] subtypePropositions = [];
+    final NonEmptyProposition[][] nonEmptyPropositions = [];
+    final Limits 'limit;
```
Comment (jclark) [inline compiler/tests/prepositionGenerator.bal]: Better to avoid the reserved word.

File: compiler/tests/prepositionGenerator.bal [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+    readonly int maxTupleMembers = 6;
+    readonly int maxStringConstLen = 20;
+    readonly int duplicationFactor = 6;
+|};
+
+type PropositionListCtor function () returns Proposition[];
+
+type OnFail function(Context cx, Proposition failedProp);
+
+class Context {
```
Comment (jclark) [inline compiler/tests/prepositionGenerator.bal]: These names are module-scoped, so I think we need a more specific name, such as GeneratorContext. Some other names here have the same problem.

**Review by jclark — CHANGES_REQUESTED** [review]

**Comment by jclark** [comment]
What tests get run automatically at the moment? Is it possible to do a longer run by hand when desired?

**Comment by rdhananjaya** [comment]
> What tests get run automatically at the moment? Is it possible to do a longer run by hand when desired?

Subtype proposition tests and Non-empty proposition tests are running automatically at the moment. Due to resource constraints in the VMs and https://github.com/ballerina-platform/ballerina-lang/issues/35923, they cannot be completed within the time limit (of 360m).

**Comment by jclark** [comment]
> they cannot be completed within the time limit

What does "they" refer to?

I don't want our normal `make test` (as run automatically by github action) to become significantly slower.

We should have some other way to run longer tests (which should be documented somewhere), which we can manually do locally when we make significant changes to semtype implementation.

**Comment by rdhananjaya** [comment]
> > they cannot be completed within the time limit
> 
> What does "they" refer to?
> 
> I don't want our normal `make test` (as run automatically by github action) to become significantly slower.
> 
> We should have some other way to run longer tests (which should be documented somewhere), which we can manually do locally when we make significant changes to semtype implementation.

"they" meant property-based tests.
Let me find a way to run those property-based tests manually without running them automatically.

File: compiler/Makefile [hunk 5]
```diff
@@ -3,10 +3,14 @@ COMPILER_JAR=../build/bin/nballerina.jar
 BAL?=bal
 BAL_BUILD_FLAGS ?= --offline
 
+ifneq (BAL_ALL_TESTS, true)
+BAL_TEST_FLAGS ?= --disable-groups longRunning
```
Comment (jclark) [inline compiler/Makefile]: I often do `make test` in the `compiler` directory. So I suggest in `compiler/Makefile` do

```
BAL_TEST_FLAGS ?= $(BAL_BUILD_FLAGS) --disable-groups longRunning

test:
        $(BAL) test $(BAL_TEST_FLAGS)
```

Then people can run all tests with `bal test`.

**Review by jclark — CHANGES_REQUESTED** [review]

**Pull request merged**